src/software/simulation/physics/physics_world.cpp

```diff
@@ -1,78 +1,211 @@
+        case TeamColour::BLUE:
+            available_func = [this](unsigned int id) {
+                return this->isBlueRobotIdAvailable(id);
+            };
+            break;
+        case TeamColour::YELLOW:
+            available_func = [this](unsigned int id) {
+                return this->isYellowRobotIdAvailable(id);
+            };
+            break;
     }
-    if (physics_field)
+
+    for (unsigned int i = 0; i < std::numeric_limits<unsigned int>::max(); i++)
+    {
+        if (available_func(i))
+        {
+            return i;
+        }
+    }
+
+    if (available_func(std::numeric_limits<unsigned int>::max()))
     {
-        new_world.mutableField() = physics_field->getField();
+        return std::numeric_limits<unsigned int>::max();
     }
 
-    new_world.mutableFriendlyTeam().clearAllRobots();
-    std::vector<Robot> friendly_robots;
-    for (const auto& robot : friendly_physics_robots)
+    LOG(FATAL) << "Out of available " << toString(colour)
+               << " robot IDs in the physics world";
+}
+
+unsigned int PhysicsWorld::getAvailableYellowRobotId() const
+{
+    return getAvailableRobotId(TeamColour::YELLOW);
+}
+
+unsigned int PhysicsWorld::getAvailableBlueRobotId() const
+{
+    return getAvailableRobotId(TeamColour::BLUE);
+}
+
+const bool PhysicsWorld::isRobotIdAvailable(unsigned int id, TeamColour colour) const
+{
+    std::vector<std::shared_ptr<PhysicsRobot>> physics_robots;
+    switch (colour)
     {
-        TimestampedRobotState timestamped_robot_state(robot->getRobotState(),
-                                                      current_timestamp);
-        friendly_robots.emplace_back(Robot(robot->getRobotId(), timestamped_robot_state));
+        case TeamColour::BLUE:
+            physics_robots = blue_physics_robots;
+            break;
+        case TeamColour ::YELLOW:
+            physics_robots = yellow_physics_robots;
+            break;
     }
-    new_world.mutableFriendlyTeam().updateRobots(friendly_robots);
 
-    new_world.mutableEnemyTeam().clearAllRobots();
-    std::vector<Robot> enemy_robots;
-    for (const auto& robot : enemy_physics_robots)
+    bool id_available = true;
+    for (const auto& robot : physics_robots)
     {
-        TimestampedRobotState timestamped_robot_state(robot->getRobotState(),
-                                                      current_timestamp);
-        enemy_robots.emplace_back(Robot(robot->getRobotId(), timestamped_robot_state));
+        if (!robot)
+        {
+            LOG(FATAL) << "Encountered a nullptr to a " << toString(colour)
+                       << " physics robot in the physics world";
+        }
+
+        if (id == robot->getRobotId())
+        {
+            id_available = false;
+            break;
+        }
     }
-    new_world.mutableEnemyTeam().updateRobots(enemy_robots);
 
-    return new_world;
+    return id_available;
+}
+
+bool PhysicsWorld::isYellowRobotIdAvailable(unsigned int id) const
```

Comment:
wondering why this function is public

Reply:
you're right it doesn't need to be. Removed these functions in favour of the private implementation (simplified some functions)